File: src/FSharp.Examples/SequenceToSequence.fs
// Copyright (c) .NET Foundation and Contributors.  All Rights Reserved.  See LICENSE in the project root for license information.
module TorchSharp.Examples.SequenceToSequence

open System
open System.IO
open System.Linq
open System.Diagnostics
open System.Collections.Generic

open TorchSharp
open TorchSharp.Modules
open type TorchSharp.torch.nn
open type TorchSharp.torch.optim

open TorchSharp.Examples

// This example is based on the PyTorch tutorial at:
//
// https://pytorch.org/tutorials/beginner/transformer_tutorial.html
//
// It relies on the WikiText2 dataset, which can be downloaded at:
//
// https://s3.amazonaws.com/research.metamind.io/wikitext/wikitext-2-v1.zip
//
// After downloading, extract the files using the defaults (Windows only).
//

let emsize = 200L
let nhidden = 200L
let nlayers = 2L
let nheads = 2L
let dropout = 0.2
let bptt = 32L

let batch_size = 64L
let eval_batch_size = 256L

let epochs = 50

let logInterval = 200

let cmdArgs = Environment.GetCommandLineArgs()

let datasetPath = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.DesktopDirectory), "..", "Downloads", "wikitext-2-v1")

torch.random.manual_seed(1L) |> ignore

let hasCUDA = torch.cuda_is_available() //torch.cuda.is_available()

let device = if hasCUDA then torch.CUDA else torch.CPU

let loss = torch.nn.CrossEntropyLoss(reduction=Reduction.Mean)
let criterion x y = loss.call(x,y)

type PositionalEncoding(dmodel, maxLen) as this =
    inherit Module<torch.Tensor,torch.Tensor>("PositionalEncoding")

    let dropout = Dropout(dropout)
    let mutable pe = torch.zeros([| maxLen; dmodel|])

    do
        let position = torch.arange(0L.ToScalar(), maxLen.ToScalar(), 1L.ToScalar()).unsqueeze(1L)
        let divTerm = (torch.arange(0L.ToScalar(), dmodel.ToScalar(), 2L.ToScalar()) * (-Math.Log(10000.0) / (float dmodel)).ToScalar()).exp()

        let NULL = System.Nullable<int64>()

        // See: https://github.com/dotnet/fsharp/issues/9369 -- for now we have to use an explicit array within the index
        //
        pe.[ [| torch.TensorIndex.Ellipsis; torch.TensorIndex.Slice(0L, NULL, 2L) |] ] <- (position * divTerm).sin()
        pe.[ [| torch.TensorIndex.Ellipsis; torch.TensorIndex.Slice(1L, NULL, 2L) |] ] <- (position * divTerm).cos()

        pe <- pe.unsqueeze(0L).transpose(0L,1L)

        this.RegisterComponents()

    override _.forward(t) =
        let NULL = System.Nullable<int64>()
        use x = t + pe.[torch.TensorIndex.Slice(NULL, t.shape.[0]), torch.TensorIndex.Slice()]
        dropout.call(x)

type TransformerModel(ntokens, device:torch.Device) as this =
    inherit Module<torch.Tensor,torch.Tensor,torch.Tensor>("Transformer")

    let pos_encoder = new PositionalEncoding(emsize, 5000L)
    let encoder_layers = TransformerEncoderLayer(emsize, nheads, nhidden, dropout)
    let transformer_encoder = TransformerEncoder(encoder_layers, nlayers)
    let encoder = Embedding(ntokens, emsize)
    let decoder = Linear(emsize, ntokens)

    let sqrEmSz = Math.Sqrt(float emsize).ToScalar()

    do
        let initrange = 0.1

        init.uniform_(encoder.weight, -initrange, initrange) |> ignore
        init.zeros_(decoder.bias) |> ignore
        init.uniform_(decoder.weight, -initrange, initrange) |> ignore

        this.RegisterComponents()

        if device.``type`` = DeviceType.CUDA then
            this.``to``(device) |> ignore

    override _.forward(t, mask) =
        let src = pos_encoder.call(encoder.call(t) * sqrEmSz)
        let enc = transformer_encoder.call(src, mask)
        decoder.call(enc)

    member _.GenerateSquareSubsequentMask(size:int64) =
        use mask = torch.ones([|size;size|]).eq(torch.tensor(1.0f)).triu().transpose(0L,1L)
        use maskIsZero = mask.eq(torch.tensor(0.0f))
        use maskIsOne = mask.eq(torch.tensor(1.0f))
        mask.to_type(torch.float32)
            .masked_fill(maskIsZero, Single.NegativeInfinity.ToScalar())
            .masked_fill(maskIsOne, 0.0f.ToScalar()).``to``(device)

let process_input (iter:string seq) (tokenizer:string->string seq) (vocab:TorchText.Vocab.Vocab) =
    torch.cat(
        [|
            for item in iter do
                let itemData = [| for token in tokenizer(item) do (int64 vocab.[token]) |]
                let t = torch.tensor(itemData)
                if t.NumberOfElements > 0L then
                    t
        |], 0L)

let batchify (data:torch.Tensor) batchSize (device:torch.Device) =
    let nbatch = data.shape.[0] / batchSize
    let d2 = data.narrow(0L, 0L, nbatch * batchSize).view(batchSize, -1L).t()
    d2.contiguous().``to``(device)

let get_batch (source:torch.Tensor) (index:int64) =

    let len = min bptt (source.shape.[0]-1L-index)
    let data = source.[torch.TensorIndex.Slice(index, index + len)]
    let target = source.[torch.TensorIndex.Slice(index + 1L, index + 1L + len)].reshape(-1L)
    data,target

let train epoch (model:TransformerModel) (optimizer:Optimizer) (trainData:torch.Tensor) ntokens =

    model.train()

    use d = torch.NewDisposeScope()

    let mutable total_loss = 0.0f
    let mutable src_mask = model.GenerateSquareSubsequentMask(bptt)

    let mutable batch = 0

    let tdlen = trainData.shape.[0]

    let mutable i = 0L

    while i < tdlen - 2L  do

        begin
            let data,targets = get_batch trainData i

            if data.shape.[0] <> bptt then
                src_mask <- model.GenerateSquareSubsequentMask(data.shape.[0])

            optimizer.zero_grad()

            let output = model.call(data, src_mask)
            let loss = criterion (output.view(-1L, ntokens)) targets
            loss.backward()
            torch.nn.utils.clip_grad_norm_(model.parameters(), 0.5) |> ignore
            optimizer.step() |> ignore

            total_loss <- total_loss + loss.cpu().item<float32>()
        end

        if (batch % logInterval = 0) && (batch > 0) then
            let cur_loss = (total_loss / (float32 logInterval)).ToString("0.00")
            printfn $"epoch: {epoch} | batch: {batch} / {tdlen/bptt} | loss: {cur_loss}"
            total_loss <- 0.0f

        batch <- batch + 1
        i <- i + bptt

        d.DisposeEverythingBut(src_mask) |> ignore


let evaluate (model:TransformerModel) (evalData:torch.Tensor) ntokens =

    model.eval()

    use d = torch.NewDisposeScope()

    let mutable total_loss = 0.0f
    let mutable src_mask = model.GenerateSquareSubsequentMask(bptt)

    let mutable batch = 0L

    let tdlen = evalData.shape.[0]

    let mutable i = 0L

    while i < tdlen - 2L  do

        begin
            let data,targets = get_batch evalData i

            if data.shape.[0] <> bptt then
                src_mask <- model.GenerateSquareSubsequentMask(data.shape.[0])

            let output = model.call(data, src_mask)
            let loss = criterion (output.view(-1L, ntokens)) targets
            total_loss <- total_loss + (float32 data.shape.[0]) * loss.cpu().item<float32>()
        end

        batch <- batch + 1L
        i <- i + bptt

        d.DisposeEverythingBut(src_mask) |> ignore

    src_mask.Dispose();

    total_loss / (float32 evalData.shape.[0])


let run epochs =

    printfn $"Running SequenceToSequence on {device.``type``.ToString()} for {epochs} epochs."

    let vocabIter = TorchText.Datasets.WikiText2("train", datasetPath)
    let tokenizer = TorchText.Data.Utils.get_tokenizer("basic_english")
    let tokenizer str = tokenizer.Invoke(str)

    let counter = new TorchText.Vocab.Counter<string>()

    for item in vocabIter do
        counter.update(tokenizer(item))

    let vocab = TorchText.Vocab.Vocab(counter)

    let trainIter,validIter,testIter = TorchText.Datasets.WikiText2(datasetPath).ToTuple()

    let train_data = batchify (process_input trainIter tokenizer vocab) batch_size device
    let valid_data = batchify (process_input validIter tokenizer vocab) eval_batch_size device
    let test_data = batchify (process_input testIter tokenizer vocab) eval_batch_size device

    let ntokens = int64 vocab.Count

    use model = new TransformerModel(ntokens, device)
    let lr = 2.50

    let optimizer = SGD(model.parameters(), lr);
    let scheduler = lr_scheduler.StepLR(optimizer, 1, 0.95, last_epoch=15)

    let totalTime = Stopwatch()
    totalTime.Start()


    for epoch = 1 to epochs do
        let sw = Stopwatch()
        sw.Start()

        train epoch model optimizer train_data ntokens

        let val_loss = evaluate model valid_data ntokens
        sw.Stop()

        let pgFirst = optimizer.ParamGroups.First()
        let lrStr = pgFirst.LearningRate.ToString("0.00")
        let elapsed = sw.Elapsed.TotalSeconds.ToString("0.0")
        let lossStr = val_loss.ToString("0.00")

        printfn $"\nEnd of epoch: {epoch} | lr: {lrStr} | time: {elapsed}s | loss: {lossStr}\n"

        scheduler.step() |> ignore

    let tst_loss = evaluate model test_data ntokens

    totalTime.Stop()

    let elapsed = totalTime.Elapsed.TotalSeconds.ToString("0.0")
    let lossStr = tst_loss.ToString("0.00")
    printfn $"\nEnd of training | time: {elapsed} s | loss: {lossStr}\n"

File: src/Examples/SequenceToSequence.cs
// Copyright (c) .NET Foundation and Contributors.  All Rights Reserved.  See LICENSE in the project root for license information.
using System;
using System.IO;
using System.Linq;
using System.Collections.Generic;
using System.Diagnostics;
using static TorchSharp.torch;
using static TorchSharp.torch.nn;

namespace TorchSharp.Examples
{

    /// <summary>
    /// This example is based on the PyTorch tutorial at:
    ///
    /// https://pytorch.org/tutorials/beginner/transformer_tutorial.html
    ///
    /// It relies on the WikiText2 dataset, which can be downloaded at:
    ///
    /// https://s3.amazonaws.com/research.metamind.io/wikitext/wikitext-2-v1.zip
    ///
    /// After downloading, extract the files using the defaults (Windows only).
    /// </summary>
    public class SequenceToSequence
    {
        // This path assumes that you're running this on Windows.
#if NET472_OR_GREATER
        private readonly static string _dataLocation = NSPath.Join(Environment.GetFolderPath(Environment.SpecialFolder.DesktopDirectory), "..", "Downloads", "wikitext-2-v1");
#else
        private readonly static string _dataLocation = Path.Join(Environment.GetFolderPath(Environment.SpecialFolder.DesktopDirectory), "..", "Downloads", "wikitext-2-v1");
#endif // NET472_OR_GREATER

        private const long emsize = 200;
        private const long nhid = 200;
        private const long nlayers = 2;
        private const long nhead = 2;
        private const double dropout = 0.2;

        private const int batch_size = 64;
        private const int eval_batch_size = 32;

        private const int epochs = 10;

        internal static void Main(string[] args)

        {
            torch.random.manual_seed(1);

            var cwd = Environment.CurrentDirectory;

            var device = torch.cuda.is_available() ? torch.CUDA : torch.CPU;
            Console.WriteLine($"Running SequenceToSequence on {device.type.ToString()} for {epochs} epochs.");

            var vocab_iter = TorchText.Datasets.WikiText2("train", _dataLocation);
            var tokenizer = TorchText.Data.Utils.get_tokenizer("basic_english");

            var counter = new TorchText.Vocab.Counter<string>();
            foreach (var item in vocab_iter) {
                counter.update(tokenizer(item));
            }

            var vocab = new TorchText.Vocab.Vocab(counter);

            var (train_iter, valid_iter, test_iter) = TorchText.Datasets.WikiText2(_dataLocation);

            var train_data = Batchify(ProcessInput(train_iter, tokenizer, vocab), batch_size).to((Device)device);
            var valid_data = Batchify(ProcessInput(valid_iter, tokenizer, vocab), eval_batch_size).to((Device)device);
            var test_data = Batchify(ProcessInput(test_iter, tokenizer, vocab), eval_batch_size).to((Device)device);

            var bptt = 32;

            var ntokens = vocab.Count;

            var model = new TransformerModel(ntokens, emsize, nhead, nhid, nlayers, dropout).to((Device)device);
            var loss = CrossEntropyLoss();
            var lr = 2.50;
            var optimizer = torch.optim.SGD(model.parameters(), lr);
            var scheduler = torch.optim.lr_scheduler.StepLR(optimizer, 1, 0.95, last_epoch: 15);

            var totalTime = new Stopwatch();
            totalTime.Start();

            foreach (var epoch in Enumerable.Range(1, epochs)) {

                var sw = new Stopwatch();
                sw.Start();

                train(epoch, train_data, model, loss, bptt, ntokens, optimizer);

                var val_loss = evaluate(valid_data, model, loss, bptt, ntokens, optimizer);
                sw.Stop();

                var pgFirst = optimizer.ParamGroups.First();
                Console.WriteLine($"\nEnd of epoch: {epoch} | lr: {pgFirst.LearningRate:0.00} | time: {sw.Elapsed.TotalSeconds:0.0}s | loss: {val_loss:0.00}\n");
                scheduler.step();
            }

            var tst_loss = evaluate(test_data, model, loss, bptt, ntokens, optimizer);
            totalTime.Stop();

            Console.WriteLine($"\nEnd of training | time: {totalTime.Elapsed.TotalSeconds:0.0}s | loss: {tst_loss:0.00}\n");
        }

        private static void train(int epoch, Tensor train_data, TransformerModel model, Loss<Tensor, Tensor, Tensor> criterion, int bptt, int ntokens, torch.optim.Optimizer optimizer)
        {
            model.train();

            using (var d = torch.NewDisposeScope()) {

                var total_loss = 0.0f;

                var batch = 0;
                var log_interval = 200;

                var src_mask = model.GenerateSquareSubsequentMask(bptt);

                var tdlen = train_data.shape[0];


                for (int i = 0; i < tdlen - 1; batch++, i += bptt) {

                    var (data, targets) = GetBatch(train_data, i, bptt);
                    optimizer.zero_grad();

                    if (data.shape[0] != bptt) {
                        src_mask = model.GenerateSquareSubsequentMask(data.shape[0]);
                    }

                    using (var output = model.call(data, src_mask)) {
                        var loss = criterion.call(output.view(-1, ntokens), targets);
                        loss.backward();
                        torch.nn.utils.clip_grad_norm_(model.parameters().ToArray(), 0.5);
                        optimizer.step();

                        total_loss += loss.to(torch.CPU).item<float>();
                    }

                    if ((batch % log_interval == 0 && batch > 0) || (batch == tdlen / bptt)) {
                        var cur_loss = total_loss / log_interval;
                        Console.WriteLine($"epoch: {epoch} | batch: {batch} / {tdlen / bptt} | loss: {cur_loss:0.00}");
                        total_loss = 0;
                    }

                    d.DisposeEverythingBut(src_mask);
                }
            }
        }

        private static double evaluate(Tensor eval_data, TransformerModel model, Loss<Tensor, Tensor, Tensor> criterion, int bptt, int ntokens, torch.optim.Optimizer optimizer)
        {
            model.eval();

            using (var d = torch.NewDisposeScope()) {

                var src_mask = model.GenerateSquareSubsequentMask(bptt);

                var total_loss = 0.0f;
                var batch = 0;


                for (int i = 0; i < eval_data.shape[0] - 1; batch++, i += bptt) {

                    var (data, targets) = GetBatch(eval_data, i, bptt);
                    if (data.shape[0] != bptt) {
                        src_mask = model.GenerateSquareSubsequentMask(data.shape[0]);
                    }
                    using (var output = model.call(data, src_mask)) {
                        var loss = criterion.call(output.view(-1, ntokens), targets);
                        total_loss += data.shape[0] * loss.to(torch.CPU).item<float>();
                    }

                    data.Dispose();
                    targets.Dispose();

                    d.DisposeEverythingBut(src_mask);
                }

                return total_loss / eval_data.shape[0];
            }
        }

        static Tensor ProcessInput(IEnumerable<string> iter, Func<string, IEnumerable<string>> tokenizer, TorchText.Vocab.Vocab vocab)
        {
            List<Tensor> data = new List<Tensor>();
            foreach (var item in iter) {
                List<long> itemData = new List<long>();
                foreach (var token in tokenizer(item)) {
                    itemData.Add(vocab[token]);
                }
                data.Add(torch.tensor(itemData.ToArray(), torch.int64));
            }

            var result = torch.cat(data.Where(t => t.NumberOfElements > 0).ToList(), 0);
            return result;
        }

        static Tensor Batchify(Tensor data, int batch_size)
        {
            var nbatch = data.shape[0] / batch_size;
            return data.narrow(0, 0, nbatch * batch_size).view(batch_size, -1).t().contiguous();
        }

        static (Tensor, Tensor) GetBatch(Tensor source, int index, int bptt)
        {
            var len = Math.Min(bptt, source.shape[0] - 1 - index);
            var data = source[TensorIndex.Slice(index, index + len)];
            var target = source[TensorIndex.Slice(index + 1, index + 1 + len)].reshape(-1);
            return (data, target);
        }

        class TransformerModel : Module<Tensor, Tensor, Tensor>
        {
            private Modules.TransformerEncoder transformer_encoder;
            private PositionalEncoding pos_encoder;
            private Modules.Embedding encoder;
            private Modules.Linear decoder;

            private long ninputs;
            private Device device;

            public TransformerModel(long ntokens, long ninputs, long nheads, long nhidden, long nlayers, double dropout = 0.5) : base("Transformer")
            {
                this.ninputs = ninputs;

                pos_encoder = new PositionalEncoding(ninputs, dropout);
                var encoder_layers = TransformerEncoderLayer(ninputs, nheads, nhidden, dropout);
                transformer_encoder = TransformerEncoder(encoder_layers, nlayers);
                encoder = Embedding(ntokens, ninputs);
                decoder = Linear(ninputs, ntokens);
                InitWeights();

                RegisterComponents();
            }

            public Tensor GenerateSquareSubsequentMask(long size)
            {
                var mask = (torch.ones(new long[] { size, size }) == 1).triu().transpose(0, 1);
                return mask.to_type(ScalarType.Float32)
                    .masked_fill(mask == 0, float.NegativeInfinity)
                    .masked_fill(mask == 1, 0.0f).to(device);
            }

            private void InitWeights()
            {
                var initrange = 0.1;

                init.uniform_(encoder.weight, -initrange, initrange);
                init.zeros_(decoder.bias);
                init.uniform_(decoder.weight, -initrange, initrange);
            }

            public override Tensor forward(Tensor t, Tensor mask)
            {
                var src = pos_encoder.call(encoder.call(t) * MathF.Sqrt(ninputs));
                var enc = transformer_encoder.call(src, mask);
                return decoder.call(enc);
            }

            protected override void Dispose(bool disposing)
            {
                if (disposing) {
                    transformer_encoder.Dispose();
                    pos_encoder.Dispose();
                    encoder.Dispose();
                    decoder.Dispose();
                    ClearModules();
                }
                base.Dispose(disposing);
            }

            protected override Module _to(DeviceType deviceType, int deviceIndex = -1)
            {
                this.device = new Device(deviceType, deviceIndex);
                return base._to(deviceType, deviceIndex);
            }
        }

        class PositionalEncoding : Module<Tensor, Tensor>
        {
            private Module<Tensor, Tensor> dropout;
            private Tensor pe;

            public PositionalEncoding(long dmodel, double dropout, int maxLen = 5000) : base("PositionalEncoding")
            {
                this.dropout = Dropout(dropout);
                var pe = torch.zeros(new long[] { maxLen, dmodel });
                var position = torch.arange(0, maxLen, 1).unsqueeze(1);
                var divTerm = (torch.arange(0, dmodel, 2) * (-Math.Log(10000.0) / dmodel)).exp();
                pe[TensorIndex.Ellipsis, TensorIndex.Slice(0, null, 2)] = (position * divTerm).sin();
                pe[TensorIndex.Ellipsis, TensorIndex.Slice(1, null, 2)] = (position * divTerm).cos();
                this.pe = pe.unsqueeze(0).transpose(0, 1);

                RegisterComponents();
            }

            public override Tensor forward(Tensor t)
            {
                using var x = t + pe[TensorIndex.Slice(null, t.shape[0]), TensorIndex.Slice()];
                return dropout.call(x);
            }

            protected override void Dispose(bool disposing)
            {
                if (disposing) {
                    dropout.Dispose();
                    pe.Dispose();
                    ClearModules();
                }
                base.Dispose(disposing);
            }
        }
    }
}
